media/server/ipc/source/WebAudioPlayerModuleService.cpp

```diff
@@ -0,0 +1,345 @@
+    }
+    catch (const std::exception &e)
+    {
+        RIALTO_SERVER_LOG_ERROR("Failed to create the web audio player module service, reason: %s", e.what());
+    }
+
+    return webAudioPlayerModule;
+}
+
+WebAudioPlayerModuleService::WebAudioPlayerModuleService(service::IWebAudioPlayerService &webAudioPlayerService)
+    : m_webAudioPlayerService{webAudioPlayerService}
+{
+}
+
+WebAudioPlayerModuleService::~WebAudioPlayerModuleService() {}
+
+void WebAudioPlayerModuleService::clientConnected(const std::shared_ptr<::firebolt::rialto::ipc::IClient> &ipcClient)
+{
+    RIALTO_SERVER_LOG_INFO("Client Connected!");
+    {
+        m_clientWebAudioPlayerHandles.emplace(ipcClient, std::set<int>());
+    }
+    ipcClient->exportService(shared_from_this());
+}
+
+void WebAudioPlayerModuleService::clientDisconnected(const std::shared_ptr<::firebolt::rialto::ipc::IClient> &ipcClient)
+{
+    RIALTO_SERVER_LOG_INFO("Client disconnected!");
+    std::set<int> webAudioPlayerHandles;
+    {
+        auto handleIter = m_clientWebAudioPlayerHandles.find(ipcClient);
+        if (handleIter == m_clientWebAudioPlayerHandles.end())
+        {
+            RIALTO_SERVER_LOG_ERROR("unknown client disconnected");
+            return;
+        }
+        webAudioPlayerHandles = handleIter->second; // copy to avoid deadlock
+        m_clientWebAudioPlayerHandles.erase(handleIter);
+    }
+    for (const auto &webAudioPlayerHandle : webAudioPlayerHandles)
+    {
+        m_webAudioPlayerService.destroyWebAudioPlayer(webAudioPlayerHandle);
+    }
+}
+
+void WebAudioPlayerModuleService::createWebAudioPlayer(::google::protobuf::RpcController *controller,
+                                                       const ::firebolt::rialto::CreateWebAudioPlayerRequest *request,
+                                                       ::firebolt::rialto::CreateWebAudioPlayerResponse *response,
+                                                       ::google::protobuf::Closure *done)
+{
+    RIALTO_SERVER_LOG_DEBUG("entry:");
+    auto ipcController = dynamic_cast<firebolt::rialto::ipc::IController *>(controller);
+    if (!ipcController)
+    {
+        RIALTO_SERVER_LOG_ERROR("ipc library provided incompatible controller object");
+        controller->SetFailed("ipc library provided incompatible controller object");
+        done->Run();
+        return;
+    }
+
+    WebAudioConfig config;
+    if (request->has_config())
+    {
+        if (request->config().has_pcm())
+        {
+            config.pcm.rate = request->config().pcm().rate();
+            config.pcm.channels = request->config().pcm().channels();
+            config.pcm.sampleSize = request->config().pcm().sample_size();
+            config.pcm.isBigEndian = request->config().pcm().is_big_endian();
+            config.pcm.isSigned = request->config().pcm().is_signed();
+            config.pcm.isFloat = request->config().pcm().is_float();
+        }
+    }
```

Comment:
What happens if there is no config or PCM parameters? For now, I suppose we should return an error in such a case.

Reply:
Yes agreed, however I will add the logic to WebAudioPlayer, not the service layer.